### ISSUE TITLE
Add option to the `no-nested-ternary` rule that allows parens wrapped nested ternaries

### ISSUE THREAD

**Issue by Qix-** [eslint/eslint#3480]
We've been discussing it over at sindresorhus/xo#12 and concluded it might be nice to have a rule where `no-nested-ternary` wouldn't trigger if the nested ternary is wrapped in parenthesis.

We agreed the following 'unreadable' and ambiguous code

```javascript
var foo = i > 5 ? true : i < 100 ? true : false;
```

could be improved by adding parenthesis.

```javascript
var foo = i > 5 ? true : (i < 100 ? true : false);
```

As it stands, ESLint catches both as nested ternaries.

--- 

There also comes the issue of logic. The following is correct logic; `platform === 'linux'` isn't evaluated unless `opts.dash` is falsey.

```javascript
var file = opts.dash ? '-' : (platform === 'linux' ? '/dev/stdin' : '/proc/self/fd/0');
```

Alternatively, the logic would have to be

```javascript
var file;
if (opts.dash) {
  file = '-';
} else {
  file = platform === 'linux' ? '/dev/stdin' : '/proc/self/fd/0';
}
```

which is much less than ideal.

---

It'd be nice if there was an option for allowing parenthesis wrapped nested ternary expressions, i.e.

```
{'no-nested-ternary': [2, {allowParensWrapped: true}]
```

// @sindresorhus

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by sindresorhus** [comment]
:+1: I'm all for enforcing readable, which nested ternaries usually aren't, but one extra level nested ternary wrapped in parens is pretty readable IMHO. So would be nice to have an exception for this.

**Comment by nzakas** [comment]
The trouble we'll run into is how this overlaps with `no-extra-parens`. Even if we add such an option, if you're also using `no-extra-parens`, it will flag those parentheses as unnecessary. We've already run into trouble with this on a couple of other rules that use parentheses as the exception to the rule. Since rules can't know about each other's settings, this puts us in a tricky situation.

**Comment by SpadeAceman** [comment]
:+1: I'm also in favor of the enhanced readability this proposal would allow.

> The trouble we'll run into is how this overlaps with `no-extra-parens`.

Looking at the `no-extra-parens` documentation, I see that the `"all"` option has exceptions baked into it, for IIFEs and the `wrap-regex` rule. Perhaps `no-extra-parens` could be modified to allow for the desired exceptions to be specifically listed?

```
// ...
,"no-extra-parens": [2, "all", {
  "exceptions": {
    "regex": true
    ,"ternary": true
    ,"assign": true // for 'if' and 'return' statements to assign values within parentheses
  }
}]
// ...
```

**Comment by nzakas** [comment]
The problem with that is that it would allow parens around all ternaries, not just nested ones.

**Comment by gyandeeps** [comment]
We left this hanging here.
ping @eslint/eslint-team

**Comment by nzakas** [comment]
Unless I'm missing something, I don't see how we can support this. If anyone has an idea, please speak now.

**Comment by SpadeAceman** [comment]
> The problem with that is that it would allow parens around all ternaries, not just nested ones.

Adding the option to `no-extra-parens` would simply open the door to allowing extra parens around ternaries. It would be up to `no-nested-ternary` (and its associated options) to enforce it around just the nested ones.

**Comment by nzakas** [comment]
I think that slices the responsibility for checking parens a bit too thin. The "functions" option for `no-extra-parens` is there for the very common case of IIFEs. I don't think this rises to that level of ubiquity.

The overlap of rules is something we try hard to avoid as it creates a lot of confusion and tight coupling between rules.

**Comment by gyandeeps** [comment]
So far I am hearing that this is not feasible.
Any updates?

**Comment by nzakas** [comment]
Yeah, I don't think we can support this. Sorry, all. 

Remember, custom rules can be made to do whatever you want.

**Comment by wmertens** [comment]
It is hard to avoid nested ternaries in React JSX expressions since it is the only available flow control inside them short of creating an immediately-evaluated function.

The only other alternative is preparing the result beforehand, which can be confusing and hard to maintain.

When doing multi-line ternaries, it also helps with figuring out nesting, somewhat.

I think that allowing parens around ternaries is an acceptable trade-off here…

```jsx
<div>{isFoo ? (isBar ? "barred foo" : <Foo />) : <Whee />}</div>
```
and multiline
```jsx
<div>{isFoo ? (
  isBar ? (
    "barred foo"
  ) : (
    <Foo />
  )
) : (
  <Whee />
)}</div>
```

**Comment by Qix-** [comment]
I still don't see why this can't be supported. This is a perfectly viable rule suggestion...

**Comment by platinumazure** [comment]
@Qix- On its own, this is a perfectly reasonable and feasible rule proposal. However, it is likely to collide with rule `no-extra-parens` and cause confusion for users. The rules can't know about each other's configuration, so users will commonly enable this proposed exception and not know to do a similar exception in `no-extra-parens` (which may not even exist at this point).

We have to be extremely careful about rule collisions here, otherwise ESLint becomes unusable or unmaintainable.

Until the ESLint team can discuss, design, and implement a sane way for rules to share this sort of information without actually coupling the rules together, this enhancement is not worth doing due to potential user confusion.

**Comment by Qix-** [comment]
Can't you detect if the parens are around a ternary operation and, if so, check to see if this rule is enabled in order to determine if an error/warning needs to be raised? Doesn't seem like such a confusing thing...

**Comment by ilyavolodin** [comment]
Qix, we don't have a mechanize to "check to see if this rule is enabled", and we don't want to couple rules among themselves, that would lead to unsustainable code. So, in short, no - rules cannot know anything about other rules.

**Comment by wmertens** [comment]
Okay, so, second best option: How would I go about making a rule that does
this? Pointers would be greatly appreciated…

On Mon, Feb 29, 2016 at 2:59 PM Ilya Volodin <notifications@github.com>
wrote:

> Qix, we don't have a mechanize to "check to see if this rule is enabled",
> and we don't want to couple rules among themselves, that would lead to
> unsustainable code. So, in short, no - rules cannot know anything about
> other rules.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eslint/eslint/issues/3480#issuecomment-190222864>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

**Comment by BYK** [comment]
@wmertens - http://eslint.org/docs/developer-guide/working-with-plugins#create-a-plugin is a good start

Then you can look at sample plugins like https://github.com/digencer/eslint-plugin-expires and adapt it. I'd suggest creating your own repo, copy the existing rule you want to modify and go over from that. Also copy the existing tests along to make sure you're not breaking anything.

If you ever need help, ping one of us in the Gitter.im chat room and we'd be happy to help! :)

**Comment by ilyavolodin** [comment]
Even easier options: https://github.com/eslint/generator-eslint has a generator to start a new plugin for ESLint. You can run that, and it will create all folders and files that you need to get started.

**Comment by SpadeAceman** [comment]
Can I request that this type of rejection be tracked for future referece? Specifically, I'm asking for some way of tagging or tracking reasonable rule or rule option requests like this, which are rejected solely because they would clash with other existing rules, in ways the ESLint team believes would be too problematic to support.

Various permutations of this issue have come up before, and will certainly continue to do so – see also parallel requests for rules to know about each others' options. Tracking these rejections would provide solid data for informed debate on this tricky subject in future.

**Comment by platinumazure** [comment]
Actually this sounds like a great idea, like a won't-fix but for a specific
reason. Something like "rule collision" maybe?

**Comment by SpadeAceman** [comment]
"rule collision" sounds good, very clear.

I've also realized this could retroactively help identify rule collision patterns over time – perhaps there's a specific rule that keeps clashing with other rules, or a pair of rules could emerge as future merge candidates after repeatedly stepping on each others' toes.

**Comment by sompylasar** [comment]
Please could anyone explain this question about the community bias: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/50#issuecomment-249417065 ?
